### PR TITLE
Naive method open mp

### PR DESCRIPTION
**Pull request by Param-29**
Updated NaiveMethod with openmp.

Serial 

```
 Performance counter stats for './a.out':

          5,503.51 msec task-clock                #    1.000 CPUs utilized          
                17      context-switches          #    0.003 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
            14,320      page-faults               #    0.003 M/sec                  
   17,00,99,78,994      cycles                    #    3.091 GHz                    
   34,69,96,64,023      instructions              #    2.04  insn per cycle         
    5,02,09,92,263      branches                  #  912.325 M/sec                  
         80,49,745      branch-misses             #    0.16% of all branches        

       5.504686467 seconds time elapsed


```

Parallel

```
 Performance counter stats for './a.out':

          5,734.69 msec task-clock                #    1.059 CPUs utilized          
                56      context-switches          #    0.010 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
            14,352      page-faults               #    0.003 M/sec                  
   17,40,57,71,386      cycles                    #    3.035 GHz                    
   33,86,87,45,017      instructions              #    1.95  insn per cycle         
    4,90,54,37,375      branches                  #  855.397 M/sec                  
         78,10,430      branch-misses             #    0.16% of all branches        

       5.414346690 seconds time elapsed

```
Speed Up of 6% for 1200 data-samples.

Other hot-spots in the code 
1)arma::eig_sym()
2)Evaluate() as i cannot find source code.

### CODE REVIEW

**Comment by zoq** [comment]
Can you share the test code, the compiler options and some system informations as well?

**Comment by Param-29** [comment]
Yes.
Following Is the testing code:

```
  arma::mat dataset;

  // Now, there are 750 points centered at the origin with unit variance.
  dataset.randn(3, 1200);
  dataset *= 0.05;

  // Take the second 250 points and spread them away from the origin.
  for (size_t i = 400; i < 800; ++i)
  {
    // Push the point away from the origin by 2.
    const double pointNorm = norm(dataset.col(i), 2);

    dataset(0, i) += 2.0 * (dataset(0, i) / pointNorm);
    dataset(1, i) += 2.0 * (dataset(1, i) / pointNorm);
    dataset(2, i) += 2.0 * (dataset(2, i) / pointNorm);
  }

  // Take the third 500 points and spread them away from the origin.
  for (size_t i = 800; i < 1200; ++i)
  {
    // Push the point away from the origin by 5.
    const double pointNorm = norm(dataset.col(i), 2);

    dataset(0, i) += 5.0 * (dataset(0, i) / pointNorm);
    dataset(1, i) += 5.0 * (dataset(1, i) / pointNorm);
    dataset(2, i) += 5.0 * (dataset(2, i) / pointNorm);
  }

  // Now we have a dataset; we will use the GaussianKernel to perform KernelPCA
  // using the naive method to take it down to one dimension.
  KernelPCA<GaussianKernel> p;
  p.Apply(dataset, 1);

  // Get the ranges of each "class".  These are all initialized as empty ranges
  // containing no points.
  Range ranges[3];
  ranges[0] = Range();
  ranges[1] = Range();
  ranges[2] = Range();
```
 It is similar to **CircleTransformationTestNystroem** in _KernelPCATest_ test suite.

Compile Time Options

g++ test.cpp -fopenmp -lmlpack -g -larmadillo

System's Information
Quad-core intel i5 7th gen Processor.

File: src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp [hunk 1]
```diff
@@ -77,7 +81,9 @@ class NaiveKernelRule
 
   // Swap the eigenvalues since they are ordered backwards (we need largest to
   // smallest).
-  for (size_t i = 0; i < floor(eigval.n_elem / 2.0); ++i)
+  omp_size_t varForOpenmp = floor(eigval.n_elem / 2.0);
+  #pragma omp parallel for
+  for (omp_size_t i = 0; i < varForOpenmp ; i++)
```
Comment (birm) [inline src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp]: Not necessarily this one, but consider renaming this variable
```suggestion
  omp_size_t num_eigenvals = floor(eigval.n_elem / 2.0);
  #pragma omp parallel for
  for (omp_size_t i = 0; i < num_eigenvals; i++)
```

File: src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp [hunk 2]
```diff
@@ -47,20 +47,24 @@ class NaiveKernelRule
...
+      kernelMatrix(j-i,i) = kernelMatrix(i, j);
     }
   }
 
   // Copy to the lower triangular part of the matrix.
+  /*
   for (size_t i = 1; i < data.n_cols; ++i)
     for (size_t j = 0; j < i; ++j)
       kernelMatrix(i, j) = kernelMatrix(j, i);
+  */
```
Comment (birm) [inline src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp]: If you're removing code, remove it :) Be bold! Git histories hold the past.

Comment (Param-29) [reply src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp]: Sure :)

**Comment by birm** [comment]
Do you need any help with the tests?

**Comment by Param-29** [comment]
> Do you need any help with the tests?

I'd like to know how do I perform these test. I can show results as I did in #2169 (using the testing code mentioned above), or I could do profiling of kernel_pca by writing some programs and checking if there are any more parts of the program that can be parallelized and sharing my results in similar way.

**Comment by birm** [comment]
It may not be a badidea to do so, but I was referring more to the fact that the test cases, namely the KernelPCAMainTest (see for example [here](https://dev.azure.com/mlpack/mlpack/_build/results?buildId=1021&view=logs&j=24d3abe3-ef0b-5deb-3aab-64d839de2c3c&t=5946fe82-49f3-5c75-2e88-cf8b07c3bc11&l=341)) This may mean that your implementation has some error.

**Comment by Param-29** [comment]
> It may not be a badidea to do so, but I was referring more to the fact that the test cases, namely the KernelPCAMainTest (see for example [here](https://dev.azure.com/mlpack/mlpack/_build/results?buildId=1021&view=logs&j=24d3abe3-ef0b-5deb-3aab-64d839de2c3c&t=5946fe82-49f3-5c75-2e88-cf8b07c3bc11&l=341)), fails. This may mean that your implementation has some error.

I missed that :). I'll work on it now and I'll ask on IRC if i have any difficulty.

File: src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp [hunk 3]
```diff
@@ -47,20 +47,20 @@ class NaiveKernelRule
...
-  for (size_t i = 0; i < data.n_cols; ++i)
+  #pragma omp parallel for
+  for (omp_size_t i = 0; i < data.n_cols; ++i)
   {
     for (size_t j = i; j < data.n_cols; ++j)
     {
       // Evaluate the kernel on these two points.
       kernelMatrix(i, j) = kernel.Evaluate(data.unsafe_col(i),
                                            data.unsafe_col(j));
+      kernelMatrix(j,i) = kernelMatrix(i, j);
```
Comment (birm) [inline src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp]:
```suggestion
      kernelMatrix(j, i) = kernelMatrix(i, j);
```

File: src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp [hunk 4]
```diff
@@ -47,20 +47,20 @@ class NaiveKernelRule
...
+  #pragma omp parallel for
+  for (omp_size_t i = 0; i < data.n_cols; ++i)
   {
     for (size_t j = i; j < data.n_cols; ++j)
     {
       // Evaluate the kernel on these two points.
       kernelMatrix(i, j) = kernel.Evaluate(data.unsafe_col(i),
                                            data.unsafe_col(j));
+      kernelMatrix(j, i) = kernelMatrix(i, j);
+      // Copy to the lower triangular part of the matrix.
```
Comment (zoq) [inline src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp]: This should be at the top of the for loop.

File: src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp [hunk 5]
```diff
@@ -77,7 +77,9 @@ class NaiveKernelRule
 
   // Swap the eigenvalues since they are ordered backwards (we need largest to
   // smallest).
-  for (size_t i = 0; i < floor(eigval.n_elem / 2.0); ++i)
+  omp_size_t num_eigenvals = floor(eigval.n_elem / 2.0);
```
Comment (zoq) [inline src/mlpack/methods/kernel_pca/kernel_rules/naive_method.hpp]: Please use camel casing for all names, see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#naming-conventions for details.

**Comment by zoq** [comment]
> Compile Time Options
> 
> g++ test.cpp -fopenmp -lmlpack -g -larmadillo
> 
> System's Information
> Quad-core intel i5 7th gen Processor.
> 
> Do you want me to compile and test with some other options? or a graph of data-points vs time similar to #2169 ?

Would be interesting to know if you tested against OpenBLAS or BLAS.

**Comment by Param-29** [comment]
> > Compile Time Options
> > g++ test.cpp -fopenmp -lmlpack -g -larmadillo
> > System's Information
> > Quad-core intel i5 7th gen Processor.
> > Do you want me to compile and test with some other options? or a graph of data-points vs time similar to #2169 ?
> 
> Would be interesting to know if you tested against OpenBLAS or BLAS.

Hey @zoq I tested it against OpenBLAS and when we use OpenMP, it kills the performance. 

Serial :
```

 Performance counter stats for './a.out':

            225.99 msec task-clock                #    0.995 CPUs utilized          
                15      context-switches          #    0.066 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
             6,359      page-faults               #    0.028 M/sec                  
      61,06,58,677      cycles                    #    2.702 GHz                    
    1,15,92,31,303      instructions              #    1.90  insn per cycle         
       6,64,60,489      branches                  #  294.084 M/sec                  
          3,29,979      branch-misses             #    0.50% of all branches        

       0.227078489 seconds time elapsed
```
Parallel:
```

 Performance counter stats for './a.out':

          1,158.88 msec task-clock                #    3.415 CPUs utilized          
               726      context-switches          #    0.626 K/sec                  
                 5      cpu-migrations            #    0.004 K/sec                  
             7,629      page-faults               #    0.007 M/sec                  
    3,29,57,13,218      cycles                    #    2.844 GHz                    
    1,87,64,23,709      instructions              #    0.57  insn per cycle         
      19,40,05,805      branches                  #  167.408 M/sec                  
         30,65,169      branch-misses             #    1.58% of all branches        

       0.339367915 seconds time elapsed

```

**Pull request closed without merging**